vmm/src/device_manager.rs

```diff
@@ -1325,88 +1328,91 @@ impl DeviceManager {
-                        ))
-                    } else {
-                        None
-                    };
-
-                    let virtio_fs_device = Arc::new(Mutex::new(
-                        vm_virtio::vhost_user::Fs::new(
-                            fs_sock,
-                            &fs_cfg.tag,
-                            fs_cfg.num_queues,
-                            fs_cfg.queue_size,
-                            cache,
-                        )
-                        .map_err(DeviceManagerError::CreateVirtioFs)?,
-                    ));
+                self._mmap_regions.push(mmap_region);
 
-                    devices.push((
-                        Arc::clone(&virtio_fs_device) as Arc<Mutex<dyn vm_virtio::VirtioDevice>>,
+                self.memory_manager
+                    .lock()
+                    .unwrap()
+                    .create_userspace_mapping(
+                        fs_guest_addr.raw_value(),
+                        fs_cache,
+                        addr,
+                        false,
                         false,
-                    ));
+                    )
+                    .map_err(DeviceManagerError::MemoryManager)?;
+
+                let mut region_list = Vec::new();
+                region_list.push(VirtioSharedMemory {
+                    offset: 0,
+                    len: fs_cache,
+                });
+
+                Some((
+                    VirtioSharedMemoryList {
+                        addr: fs_guest_addr,
+                        len: fs_cache as GuestUsize,
+                        region_list,
+                    },
+                    addr,
+                ))
+            } else {
+                None
+            };
+
+            let virtio_fs_device = Arc::new(Mutex::new(
+                vm_virtio::vhost_user::Fs::new(
+                    fs_sock,
+                    &fs_cfg.tag,
+                    fs_cfg.num_queues,
+                    fs_cfg.queue_size,
+                    cache,
+                )
+                .map_err(DeviceManagerError::CreateVirtioFs)?,
+            ));
+
+            self.add_migratable_device(Arc::clone(&virtio_fs_device) as Arc<Mutex<dyn Migratable>>);
+
+            return Ok((Arc::clone(&virtio_fs_device) as VirtioDeviceArc, false));
+        }
 
-                    let migratable = Arc::clone(&virtio_fs_device) as Arc<Mutex<dyn Migratable>>;
-                    let id = migratable.lock().unwrap().id();
-                    self.migratable_devices.insert(id, migratable);
+        Err(DeviceManagerError::NoVirtioFsSock)
+    }
+
+    fn make_virtio_fs_devices(&mut self) -> DeviceManagerResult<Vec<(VirtioDeviceArc, bool)>> {
+        let mut devices = Vec::new();
+
+        let fs_devices = self.config.lock().unwrap().fs.clone();
+        if let Some(fs_list_cfg) = &fs_devices {
+            for fs_cfg in fs_list_cfg.iter() {
+                if let Some(_) = fs_cfg.sock.to_str() {
```

Comment:
You don't need this line as it is checked from inside `make_virtio_fs_device()` directly.